src/ipld/util.rs

```diff
@@ -353,27 +377,61 @@ impl<DB: Blockstore, T: Iterator<Item = Tipset> + Unpin> Stream for ChainStream<
+                    Prefetch(_) => {
+                        if let Some(Prefetch(cid_vec)) = this.dfs.pop_front() {
+                            let cid_vec_prefetch = cid_vec.clone();
+                            this.dfs.push_front(Iterate(cid_vec));
+
+                            let cur_count = this.prefetch_worker_count.load(Ordering::Relaxed);
+                            // Limit the amount of prefetching tasks in order to maximize
+                            // performance.
+                            if cur_count < *this.max_prefetch_worker_count {
+                                task::spawn({
+                                    let db = this.db.clone();
+                                    let count = this.prefetch_worker_count.clone();
+                                    let seen = this.seen.clone();
+                                    async move {
+                                        for cid in cid_vec_prefetch {
+                                            if seen.read().exists(cid) {
+                                                continue;
+                                            }
+                                            let _ = db.get(&cid);
                                         }
-                                        return Poll::Ready(Some(Ok(Block { cid, data })));
-                                    } else if *this.fail_on_dead_links {
-                                        return Poll::Ready(Some(Err(anyhow::anyhow!(
-                                            "missing key: {}",
-                                            cid
-                                        ))));
+                                        count.fetch_sub(1, Ordering::Relaxed);
+                                    }
+                                });
+                            }
+                        }
+                    }
+                    Iterate(cid_vec) => {
+                        while let Some(cid) = cid_vec.pop_front() {
+                            // The link traversal implementation assumes there are three types of encoding:
+                            // 1. DAG_CBOR: needs to be reachable, so we add it to the queue and load.
+                            // 2. IPLD_RAW: WASM blocks, for example. Need to be loaded, but not traversed.
+                            // 3. _: ignore all other links
+                            // Don't revisit what's already been visited.
+                            if should_save_block_to_snapshot(cid) && this.seen.write().insert(cid) {
+                                if let Some(data) = this.db.get(&cid)? {
+                                    if cid.codec() == fvm_ipld_encoding::DAG_CBOR {
+                                        let ipld: Ipld = from_slice_with_fallback(&data)?;
+                                        let mut new_vec = DfsIter::new(ipld)
+                                            .flat_map(ipld_to_cid)
```

Comment:
I'd feel more comfortable with `filter_map` here. I'm not 100% sure if `flat_map` optimizes as well.

Reply:
You are correct, `filter_map` most likely has better performance.